### ISSUE TITLE
Why inductive?

### ISSUE THREAD

**Issue by Mrslock** [GraphSAINT/GraphSAINT#11]
Hello, thanks for sharing your code! In readme.md, you said "For simplicity of implementation, during validation / test set evaluation, we perform layer propagation using the full graph adjacency matrix.", So when a new node come, all the node embeddings need to be calculated, but for example, in GraphSAGE, when a new node come, not all embeddings but some nodes are involved which I think is what "inductive" truely means. In other words, I suppose a "inductive" graph learning method should:1. can be trained without val and test nodes and links;2. can generate node embedding for a new comming node very fast. It seems to me that graphSAINT can satisfy 1 but can not satisfy 2. Is that correct? Thanks!

**Comment by ZimpleX** [comment]
Thanks for the question. These two points are good summaries. To elaborate it further, point 1 is a requirement to be fulfilled by the *training* algorithm, while point 2 is a requirement to be satisfied by the *inference* algorithm. The inference and training algorithms can be designed independently and may not directly relate to each other. Consider the following scenario under a L-layer GCN model:

When a new node v comes, GraphSAINT and any of the baselines (i.e., vanilla GCN, GraphSAGE, S-GCN, AS-GCN, FastGCN, ClusterGCN) can generate the embedding (or class label) of v, by aggregating information of neighbors up to L-hop (i.e., not the neighbors spanning the full graph). Therefore, all these 7 methods can satisfy your requirement 2 using the same inference algorithm, even though they follow quite different training procedure. 

If you want to further speedup the inference, you can perform sampling on the multi-hop neighbors of v, just like what GraphSAGE did in their original paper. Note that such neighbor sampling *during inference* is still applicable to all the various training methods (including GraphSAINT). 

The scope of GraphSAINT is to improve the quality of training, not inference. Also, to make fair comparison, when evaluating the test/validation set accuracy in our paper, we don't perform sampling in the inference phase for GraphSAINT or any of the baselines. 

*Question*: So when calculating the val/test set accuracy, why do we use the adj of the full graph and then mask the relevant node?

*Answer*: It is more of a design choice to ease implementation. If we were to perform inference using the algorithm I described above, we would have to implement another function to grab/sample all the neighbor nodes up to L-hop. We would also have to modify the computation graph (now we cannot use the same adj across different layers, since the k-hop neighbors are not the same as (k+1)-hop neighbors). So these modifications are a bit tricky to handle during coding. In addition, note that the sizes of the validation/test sets are large enough (~30% of the total nodes). Even if you were to implement the neighbor-sampling based inference, you would end up visiting the full graph (2-hop neighbors of all the val/test nodes are likely to cover >90% of the full graph). So there really is not much saving in computation --- Again, this conclusion on computation efficiency during inference is true for GraphSAGE and also any other baselines. This is mainly due to the fact that there are lots of val/test nodes. 

Lastly, I would like to point out that whether you perform inference in the L-hop neighbor-sampling fashion or in our full graph fashion, you get exactly the same embedding / classification values.

**Comment by Mrslock** [comment]
So, GraphSAINT is designed to train a GCN quickly and precisely, this is the key point. While during test/inference, for ease of implementaion, we can use the full batch training method as we use for other baselines when doing test/inference.
I think I get your point. Thx again!

**Comment by ZimpleX** [comment]
That's correct. 

You're welcome!
